modules/dnn/src/cuda4dnn/csl/cudnn/cudnn.hpp

```diff
@@ -0,0 +1,201 @@
+        TensorDescriptor(TensorDescriptor&& other) noexcept
+            : descriptor{ other.descriptor } {
+            other.descriptor = nullptr;
+        }
+
+        /** constructs a tensor descriptor from the axis lengths provided in \p shape */
+        template <class SequenceContainer, typename = decltype(std::begin(std::declval<SequenceContainer>()))>
+        TensorDescriptor(const SequenceContainer& shape) {
+            constructor(shape.begin(), shape.end());
+        }
+
+        /** constructs a tensor descriptor from the axis lengths provided in [begin, end) */
+        template <class ForwardItr, typename = typename std::enable_if<!std::is_integral<ForwardItr>::value, void>::type> // TODO is_iterator
+        TensorDescriptor(ForwardItr begin, ForwardItr end) {
+            constructor(begin, end);
+        }
+
+        /** constructs a tensor descriptor from the axis lengths provided as arguments */
+        template <class ...Sizes>
+        TensorDescriptor(Sizes ...sizes) {
+            static_assert(sizeof...(Sizes) <= CUDNN_DIM_MAX, "required rank exceeds maximum supported rank");
+            std::array<int, sizeof...(Sizes)> dims = { static_cast<int>(sizes)... };
+            constructor(std::begin(dims), std::end(dims));
+        }
+
+        ~TensorDescriptor() noexcept {
+            if (descriptor != nullptr) {
+                /* cudnnDestroyTensorDescriptor will not fail */
+                CUDA4DNN_CHECK_CUDNN(cudnnDestroyTensorDescriptor(descriptor));
+            }
+        }
+
+        TensorDescriptor& operator=(const TensorDescriptor&) = delete;
+        TensorDescriptor& operator=(TensorDescriptor&& other) noexcept {
+            descriptor = other.descriptor;
+            other.descriptor = nullptr;
+            return *this;
+        };
+
+        cudnnTensorDescriptor_t get() const noexcept { return descriptor; }
+
+    private:
+        template <class ForwardItr>
+        void constructor(ForwardItr start, ForwardItr end) {
+            CV_Assert(start != end);
+            CV_Assert(std::distance(start, end) <= CUDNN_DIM_MAX);
+
+            CUDA4DNN_CHECK_CUDNN(cudnnCreateTensorDescriptor(&descriptor));
+            try {
+                const auto rank = std::distance(start, end);
+                if (rank <= 4) {
+                    std::array<int, 4> dims;
+                    std::fill(std::begin(dims), std::end(dims), 1);
+
+                    /* suppose we have a 3d tensor, the first axis is the batch axis and
+                     * the second axis is the channel axis (generally)
+                     *
+                     * cuDNN frequently assumes that the first axis is the batch axis and the
+                     * second axis is the channel axis; hence, we copy the shape of a lower rank
+                     * tensor to the begining of `dims`
+                     */
+                    std::copy(start, end, std::begin(dims));
```

Comment:
I have doubts over the validity of this 3d to 4d tensor conversion. It looks alright but I'm not very confident.

If a tensor has a shape of `[b, c, w]`, a 4D tensor descriptor with the shape `[b, c, w, 1]` is created. 

cuDNN documentation suggests using the 4d tensor API over Nd tensor API whenever possible. Hence, this weird conversion code is present even though the Nd tensor API can create 3d tensors.

Reply:
Why do you have doubts about it?  In and of itself it’s a fine thing to do. So long as code that uses it understands that there is this interpretation that the first channel is the batch, then the num_channels, then width, then height, then it’s fine to auto set unspecified dimensions to 1. 

You should make this very clear in comments at the top of this class though.

Reply:
I think cuDNN expects 3d tensors for LSTMs. LSTMs are not yet supported by the backend but it might cause trouble while attempting to implement it.

Reply:
You could presumably do the appropriate conversion in those cases though, since if you have to call an API that needs a rank 3 tensor, you can check if the rank 4 tensor is effectively rank 3 (it's last dim is 1) and then do the right thing there.  So it doesn't seem like what you are doing here will be a huge deal if LSTMs are eventually supported.